File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
TreeNode* solve(vector<int>& preorder, vector<int>& inorder, int inStart,int inEnd, int preStart, int preEnd){
    if(inStart>inEnd) return NULL;
    TreeNode* root= new TreeNode(preorder[preStart]);
    int i=inStart;
    for(;i<=inEnd;i++){
        if(inorder[i]==root->val) break;
    }
    int leftSize=i-inStart;
    int rightSize= inEnd-i;
    root->left= solve(preorder,inorder, inStart,i-1,preStart+1, preStart+leftSize);
    root->right= solve(preorder,inorder, i+1,inEnd,preEnd-rightSize+1, preEnd);
    return root;

}
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int n=preorder.size();
        int inStart=0,inEnd=n-1;
        int preStart=0, preEnd=n-1;
        return solve(preorder,inorder, inStart, inEnd, preStart, preEnd);
    }
};

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function(preorder, inorder) {
     const n=preorder.length;
    function solve(preorder,  inorder, inStart,inEnd, preStart, preEnd){
    if(inStart>inEnd) return null;
    const root= new TreeNode(preorder[preStart]);
    let i=inStart;
    for(;i<=inEnd;i++){
        if(inorder[i]==root.val) break;
    }
    let leftSize=i-inStart;
    let rightSize= inEnd-i;
    root.left= solve(preorder,inorder, inStart,i-1,preStart+1, preStart+leftSize);
    root.right= solve(preorder,inorder, i+1,inEnd,preEnd-rightSize+1, preEnd);
    return root;

}
    
       
        const inStart=0,inEnd=n-1;
        const preStart=0, preEnd=n-1;
        return solve(preorder,inorder, inStart, inEnd, preStart, preEnd);

};